osu.Game/Input/RealmKeyBindingStore.cs

```diff
@@ -0,0 +1,154 @@
+        public List<RealmKeyBinding> Query<T>(T action)
+            where T : Enum
+        {
+            int lookup = (int)(object)action;
+
+            return query(null, null).Where(rkb => rkb.Action == lookup).ToList();
+        }
+
+        /// <summary>
+        /// Update the database mapping for the provided key binding.
+        /// </summary>
+        /// <param name="keyBinding">The key binding to update. Can be detached from the database.</param>
+        /// <param name="modification">The modification to apply to the key binding.</param>
+        public void Update(IHasGuidPrimaryKey keyBinding, Action<IKeyBinding> modification)
+        {
+            // the incoming instance could already be a live access object.
+            Live<RealmKeyBinding>? realmBinding = keyBinding as Live<RealmKeyBinding>;
+
+            using (var realm = ContextFactory.GetForWrite())
+            {
+                if (realmBinding == null)
+                {
+                    // the incoming instance could be a raw realm object.
+                    if (!(keyBinding is RealmKeyBinding rkb))
+                        // if neither of the above cases succeeded, retrieve a realm object for further processing.
+                        rkb = realm.Context.Find<RealmKeyBinding>(keyBinding.ID);
+
+                    realmBinding = new Live<RealmKeyBinding>(rkb, ContextFactory);
+                }
+
+                realmBinding.PerformUpdate(modification);
+            }
+
+            KeyBindingChanged?.Invoke();
+        }
+
+        private void insertDefaults(IEnumerable<IKeyBinding> defaults, int? rulesetId = null, int? variant = null)
+        {
+            using (var usage = ContextFactory.GetForWrite())
+            {
+                // compare counts in database vs defaults
+                foreach (var group in defaults.GroupBy(k => k.Action))
+                {
+                    int count = query(rulesetId, variant).Count(k => k.Action == (int)group.Key);
+                    int aimCount = group.Count();
+
+                    if (aimCount <= count)
+                        continue;
+
+                    foreach (var insertable in group.Skip(count).Take(aimCount - count))
+                    {
+                        // insert any defaults which are missing.
+                        usage.Context.Add(new RealmKeyBinding
+                        {
+                            ID = Guid.NewGuid().ToString(),
+                            KeyCombination = insertable.KeyCombination.ToString(),
+                            Action = (int)insertable.Action,
+                            RulesetID = rulesetId,
+                            Variant = variant
+                        });
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Retrieve live queryable <see cref="RealmKeyBinding"/>s for a specified ruleset/variant content.
```

Comment:
are these really live? i'd expect something that's not a `Live<T>` cannot be live?

Reply:
An `IQueryable` is running native queries on the database (ie. not linq), and *all realm objects* are 100% live by nature (that is the main selling point of Realm). As in each property is read from disk, on each access.

Reply:
If that's the case then I'm slightly confused as to what the purpose of `Live<T>` actually is, because it's not used in this context (and it seems that it would be appropriate to do so).

I'm also not sure how the callers of `query` that claim to be returning detached entities actually detach them - they're not calling `Detach()` as far as I can see, and the `ToList()` that they *are* calling doesn't detach, right?

Reply:
If that's the case then I'm slightly confused as to what the purpose of `Live<T>` actually is, because it's not used in this context (and it seems that it would be appropriate to do so).

I'm also not sure how the callers of `query` that claim to be returning detached entities actually detach them - they're not calling `Detach()` as far as I can see, and the `ToList()` that they *are* calling doesn't detach, right?

Reply:
The xmldoc mentioning detaching there was out of date (i'm doing the detach at the point of usage now), oops.